cmd/state-svc/internal/server/server.go

```diff
@@ -55,18 +64,43 @@ func (s *Server) Port() int {
 	return s.port
 }
 
+func (s *Server) Wait() {
+	<-s.done
+}
+
```

Comment:
Note: This presents a sort of crossroads of error handling. The errors encountered in the lifetime of Server could relate to 1. startup errors 2. runtime errors 3. shutdown errors. Currently, the start method returns the start errors. The runtime errors are likely all coming from start as well. And the shutdown errors are logged in start.

I suggest that start should be focused on start errors, wait should be focused on runtime errors (though, we may not even be tracking/handling any "runtime errors"), and shutdown errors should be exposed by the shutdown method being exported. The behavior in close and shutdown should probably be in one function, but I may misunderstand a need to separate them. If this is not as simple as a bit of code shuffling, it might be worth ignoring this comment.

Reply:
I hear you.  The reason I do it this way, is: When we issue the shutdown request on the `/_quit` route, we still want that route to return with a 200 response, indicating: "The server is shutting down now".  So, we have to run the `Shutdown` in a go routine, but also wait for that go routine to return.  If we don't it was never able to terminate, because the main function returned before that.

So, in my first approach, I put that go routine (the one shutting down the server) in the state-svc main function, as that is the place to wait for it.  In this iteration, that shutdown-go function moved to the `server`, which as you are now rightly pointing out, leads to a very convoluted interface for a server.

Thinking about it this way, I believe that I should merge the two approaches:
- do all shutdown handling in the main routine (as in the first commit), but
- still use a context for server cancellation
- and just forward the cancel function to the server route, such that we can cancel the *global* context in the `/_quit` route.  

Does that sound correct?

Reply:
Of course, we could also go a different route, and in the `/_quit` route resolver, do the actual server shutdown (not in a go-routine), but as a consequence that route would never return, and the client issuing that `/_quit` request would have to expect a `connection interrupted` error as the only successful outcome.  That does not feel like the the correct approach though...

Reply:
I agree that 200 responses make sense to indicate the shutdown request was successful. I stand by my original suggestion to achieve any improvement so long as it's just shuffling a little code from here. It's clear you understand where I'm coming from and while I haven't dug deep into your comment above, I'm happy to trust your suggestion.